### PR TITLE
OMEMO encryption

### PR DESCRIPTION
**Pull request by vanitasvitae**
I cleaned the code up, fixed all gradle checkstyle issues and targeted 4.2

### CODE REVIEW

**Comment by vanitasvitae** [comment]
This PR contains #116

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: be2af1f52098b9f03e8838aea9d565753bbfa4e1
  * Your subject line is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by mnemonicflow** [comment]
@vanitasvitae What options are there for someone who would like to use the [XEP-0384: OMEMO Encryption](https://xmpp.org/extensions/xep-0384.html) capabilities of Smack without the GPLv3 licensed dependency -- smack-omemo-signal?

In the docs [Encrypting messages with OMEMO](https://github.com/vanitasvitae/smack-omemo/blob/master/documentation/omemo.md) I read that: 

> It is also possible, to port smack-omemo to other libraries implementing the double ratchet algorithm.

Can you give any hints as to what needs to be implemented for such a port?
You would need to have an implementation similar to smack-omemo-signal or do you also need to change code inside smack-omemo?

The XMPP OMEMO spec says:
> This XEP defines a protocol that leverages Olm [3] encryption to provide multi-end to multi-end encryption, allowing messages to be synchronized securely across multiple clients, even if some of them are offline. Olm is a cryptographic double ratched protocol based on work by Trevor Perrin and Moxie Marlinspike first published as the Axolotl protocol. 
> 
> OMEMO currently uses version 1 Olm protocol. Instead of an Axolotl key server, Personal Eventing Protocol (XEP-0163) [6] (PEP) is used to publish key data. 

Suggesting that an OMEMO implementation should use Olm (v1)?

[Matrix](https://en.wikipedia.org/wiki/Matrix_(communication_protocol)) provides end-to-end encryption via the [Olm](https://matrix.org/docs/spec/olm.html) and [MegOlm](https://matrix.org/docs/spec/megolm.html) cryptographic ratchets, which are implementations of the [Double Ratchet cryptographic ratchet](https://en.wikipedia.org/wiki/Double_Ratchet_Algorithm) in C++: https://matrix.org/git/olm/tree/, licensed under Apache License 2.0

If there was a Java implementation for Olm and MegaOlm could this be used instead of [libsignal-protocol-java](https://github.com/WhisperSystems/libsignal-protocol-java) ?

For more info about Olm see (starting from slide 23 onwards):
- [Slides - Encrypting Matrix: Building a universal end-to-end encrypted communication ecosystem with Matrix and Olm](https://matrix.org/blog/wp-content/uploads/2017/02/2017-02-03.1-FOSDEM-Encrypting-Matrix.pdf)
 - [Video - Encrypting Matrix: Building a universal end-to-end encrypted communication ecosystem with Matrix and Olm](http://bofh.nikhef.nl/events/FOSDEM/2017/AW1.125/matrix_future.vp8.webm)
- [Matrix’s ‘Olm’ End-to-end Encryption security assessment released – and implemented cross-platform on Riot at last!](https://matrix.org/blog/2016/11/21/matrixs-olm-end-to-end-encryption-security-assessment-released-and-implemented-cross-platform-on-riot-at-last/)
 - [Video - The future of decentralised communication, identity and reputation with Matrix](https://fosdem.cu.be/2017/AW1.125/matrix_future.vp8.webm)

**Comment by vanitasvitae** [comment]
@mnemonicflow First of all, a non-GPL client would have to use a library that provides the same ratcheting capabilities as libsignal. I think it also has to use the same elliptic curve parameters.

Since all other implementations of OMEMO are using libsignal at the moment, I guess it isn't trivial to create a non-signal module that is fully compatible.

Olm needs to be patched in order to be interoperable with libsignal I guess. I heard they're using some different crypto parameters and all in all another wire format, but I haven't looked into it.

The OMEMO XEP is in the process of moving away from Olm as ratchet basis. Instead the goal is to define the crypto library independent (ODR = OMEMO Double Ratchet), so that it will be possible for non-GPL clients to use another library that is compatible in the future. Take a look at https://github.com/xsf/xeps/pull/460 for more information. This is the reason why I split smack-omemo and smack-omemo-signal in the first place.

If you take a look at smack-omemo-signal, you see that there is not much complicated code there, so it should be pretty straight forward to write smack-omemo-XYZ, provided library XYZ has a similar form to libsignal.

TL;DR
Olm must be more or less heavily patched for compatibility, but non-signal libraries should be possible (provided someone is willing to write them ;P). Writing smack-omemo-XYZ is the least complicated part of the work.

**Comment by mnemonicflow** [comment]
Does Smack support [XEP-0334: Message Processing Hints](https://xmpp.org/extensions/xep-0334.html) ?

I couldn't find it here:
 - [Supported extensions](https://github.com/igniterealtime/Smack/blob/master/documentation/extensions/index.md)
 - [Extensions docs](https://github.com/igniterealtime/Smack/tree/master/documentation/extensions)

As the [latest XEP-0384 OMEMO encryption specification](https://cdn.rawgit.com/mnemonicflow/6fa778086b4fcadafe57118f79b84c43/raw/b13faa72055c5dc2ce1e13068a3adc6a82d83de6/xep-0384.html) says this:

> As the asynchronous nature of OMEMO enables currently offline devices to decrypt received messages at a later time, clients MUST include a Message Processing Hints (XEP-0334) [9] <store /> hint in their OMEMO messages.   
Otherwise, server implementations of Message Archive Management (XEP-0313) [6] will generally not retain OMEMO messages, since they do not contain a `<body /> `

**Comment by vanitasvitae** [comment]
Isn't this a thing **clients** have to support?

File: smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/OmemoManager.java [hunk 1]
```diff
@@ -289,6 +289,9 @@ public static void addExplicitMessageEncryptionHint(Message omemoMessage) {
      * @throws SmackException.NoResponseException   wrong
      */
     public boolean resourceSupportsOmemo(Jid resource) throws XMPPException.XMPPErrorException, SmackException.NotConnectedException, InterruptedException, SmackException.NoResponseException {
```
Comment (Flowdalic) [inline smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/OmemoManager.java]: This should probably be a FullJid then. And I'd name it fullJid, not 'resource'.

Comment (vanitasvitae) [inline smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/OmemoManager.java]: ah, fulljid is what i thought jid would be :D

Comment (vanitasvitae) [inline smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/OmemoManager.java]: Fixed it

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: c636e72a9d3f0327005cd0b3dacbc264f268b3d9
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by Flowdalic** [comment]
I've pushed support for Hints and EME into the 4.2 branch. You may want to merge. :)

File: smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/FileBasedOmemoStore.java [hunk 2]
```diff
@@ -676,8 +676,9 @@ public CachedDeviceList loadCachedDeviceList(BareJid contact) {
             String[] ids = a.split(",");
 
             for (String id : ids) {
-                if(!id.isEmpty())
+                if(!id.trim().equals("")) {
```
Comment (Flowdalic) [inline smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/FileBasedOmemoStore.java]: nit: why not `trim().isEmpty()`?

Comment (vanitasvitae) [inline smack-omemo/src/main/java/org/jivesoftware/smackx/omemo/FileBasedOmemoStore.java]: I replaced `isEmpty` with `trim().equals("")` because gradle told me, that it doesn't know `isEmpty()` when I ran `gradle check`

**Comment by Flowdalic** [comment]
Just noticed: Could you s/elements/element/ in the package names?

**Comment by vanitasvitae** [comment]
@Flowdalic Done :)

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 687cada28613d7f70b9c37c39ae9c8845b5e5557
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 687cada28613d7f70b9c37c39ae9c8845b5e5557
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 687cada28613d7f70b9c37c39ae9c8845b5e5557
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/igniterealtime/Smack/wiki/Guidelines-for-Smack-Developers-and-Contributors

---
This message was auto-generated by https://gitcop.com
